File: scratch/parse_braces.js
const fs = require('fs');

const fileContent = fs.readFileSync('frontend/src/pages/admin/AddProductWizard.tsx', 'utf8');
const lines = fileContent.split('\n');

const targetLine = 2157;
// Reconstruct content starting from line 2080 (0-indexed line 2079)
const startIdx = fileContent.indexOf(lines[targetLine - 1]);
const content = fileContent.slice(startIdx);

let braceCount = 0;
let parenCount = 0;
let inString = false;
let stringChar = null;
let escape = false;

for (let idx = 0; idx < content.length; idx++) {
  const char = content[idx];
  if (escape) {
    escape = false;
    continue;
  }
  if (char === '\\') {
    escape = true;
    continue;
  }
  if (inString) {
    if (char === stringChar) {
      inString = false;
    }
    continue;
  }
  if (char === '"' || char === "'" || char === '`') {
    inString = true;
    stringChar = char;
    continue;
  }

  if (char === '{') {
    braceCount++;
  } else if (char === '}') {
    braceCount--;
    if (braceCount === 0) {
      const newLinesCount = content.slice(0, idx).split('\n').length - 1;
      const matchedLineIdx = targetLine - 1 + newLinesCount;
      console.log(`Brace closed at relative char index ${idx}, line ${matchedLineIdx + 1}`);
      console.log("Surrounding lines:");
      for (let i = Math.max(0, matchedLineIdx - 5); i <= Math.min(lines.length - 1, matchedLineIdx + 5); i++) {
        console.log(`${i + 1}: ${lines[i]}`);
      }
      break;
    }
  }
}

File: scratch/parse_braces.py
with open('frontend/src/pages/admin/AddProductWizard.tsx', 'r', encoding='utf-8') as f:
    lines = f.readlines()

target_line = 2157
content = "".join(lines[target_line-1:])
# We want to find the matching brace/parenthesis for the `{` at the start of line 2080
# Let's count braces and parentheses starting from the `{`
brace_count = 0
paren_count = 0
in_string = False
string_char = None
escape = False

for idx, char in enumerate(content):
    if escape:
        escape = False
        continue
    if char == '\\':
        escape = True
        continue
    if in_string:
        if char == string_char:
            in_string = False
        continue
    if char in ['"', "'", '`']:
        in_string = True
        string_char = char
        continue
    
    if char == '{':
        brace_count += 1
    elif char == '}':
        brace_count -= 1
        if brace_count == 0:
            print(f"Brace closed at relative char index {idx}, line {target_line + content[:idx].count(chr(10))}")
            # Print surrounding content
            print("Surrounding lines:")
            matched_line_idx = target_line - 1 + content[:idx].count('\n')
            for i in range(max(0, matched_line_idx - 5), min(len(lines), matched_line_idx + 6)):
                print(f"{i+1}: {lines[i]}", end="")
            break
